### PR TITLE
Elasticsearch configuration strategy support

### PR DESCRIPTION
**Pull request by kuisathaverat**
## Description

The current implementation uses `autodetect` configuration strategy, this strategy is fine in most cases. This PR allows to choose the configuration strategy from a set of pre-configured strategies.

## Motivation and Context

In our case we use SNAPSHOT versions and this strategy does not work as expected when we changed the Docker image for a new SNAPSHOT version, we need to perform a `rolling` with `groupBy` to `__all__` to perform a stop and start of all nodes at the same time to avoid version conflicts.

## How Has This Been Tested?

I have made a manual test and the PR includes some unit tests.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improves code quality but has no user-facing effect)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation

## Readiness Checklist

- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: docs/resources/ec_deployment.md [hunk 1]
```diff
@@ -349,6 +380,16 @@ The optional `elasticsearch.trust_external` block, allows external trust relatio
 * `trust_all` (Optional) If true, all clusters in this external entity will be trusted and the `trust_allowlist` is ignored.
 * `trust_allowlist` (Optional) The list of clusters to trust. Only used when `trust_all` is `false`.
 
+##### Strategy
+
+The optional `elasticsearch.strategy` allow to choose the configuration strategy used to apply the changes.
```
Comment (tobio) [inline docs/resources/ec_deployment.md:385]: IMO we should have something in here along the lines of "don't change this unless you know that you need to", it's not functionality customers should be concerned with.

File: ec/ecresource/deploymentresource/schema_elasticsearch.go [hunk 2]
```diff
@@ -385,11 +387,10 @@ func newSnapshotSourceSettings() *schema.Schema {
 
 func newExtensionSchema() *schema.Schema {
 	return &schema.Schema{
-		Type:        schema.TypeSet,
-		Set:         esExtensionHash,
+		Type:        schema.TypeList,
 		Description: "Optional Elasticsearch extensions such as custom bundles or plugins.",
 		Optional:    true,
-		MinItems:    1,
+		MaxItems:    1,
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply ec/ecresource/deploymentresource/schema_elasticsearch.go:393]: I will revert it, is a mistake change.

File: docs/resources/ec_deployment.md (hunk 1 above)
Comment (kuisathaverat) [reply docs/resources/ec_deployment.md]: I have added kind of a warning.

File: ec/ecresource/deploymentresource/elasticsearch_expanders.go [hunk 3]
```diff
@@ -135,9 +144,44 @@ func expandEsResource(raw interface{}, res *models.ElasticsearchPayload) (*model
...
+		} else if rawValue == rollingGrowAndShrink {
+			strategy.RollingGrowAndShrink = new(models.RollingGrowShrinkStrategyConfig)
+		} else if rawValue == rollingAll {
+			strategy.Rolling = &models.RollingStrategyConfig{
+				GroupBy: "__all__",
+			}
+		}
+	}
+	return res
+}
```
Comment (tobio) [inline ec/ecresource/deploymentresource/elasticsearch_expanders.go:183]: IMO we should remove the return values entirely here. If we're going to keep returning something, then we should probably use the return value [here](https://github.com/elastic/terraform-provider-ec/pull/507/files#diff-ad781c304fc98e99326fac3c762954eda56bb8f7082fe00d7522bf6e43b8570dR154)

 ```suggestion
func expandStrategy(raw interface{}, strategy *models.PlanStrategy) {
	for _, rawStrategy := range raw.([]interface{}) {
		strategyCfg, ok := rawStrategy.(map[string]interface{})
		if !ok {
			continue
		}
		rawValue := strategyCfg["type"].(string)
		if rawValue == autodetect {
			strategy.Autodetect = new(models.AutodetectStrategyConfig)
		} else if rawValue == growAndShrink {
			strategy.GrowAndShrink = new(models.GrowShrinkStrategyConfig)
		} else if rawValue == rollingGrowAndShrink {
			strategy.RollingGrowAndShrink = new(models.RollingGrowShrinkStrategyConfig)
		} else if rawValue == rollingAll {
			strategy.Rolling = &models.RollingStrategyConfig{
				GroupBy: "__all__",
			}
		}
	}
}
```

**Review by tobio — COMMENTED** [review]
Code lgtm, sorry for the back and forth I thought about this a little more last night.

What's the behaviour here when the current plan strategy doesn't match the strategy in the TF plan? We shouldn't trigger any deployment changes as a result of a strategy change, we may need to define `DiffSuppressFunc` on the `strategyResourceSchema`.

File: docs/resources/ec_deployment.md [hunk 4]
```diff
@@ -349,6 +380,16 @@ The optional `elasticsearch.trust_external` block, allows external trust relatio
 * `trust_all` (Optional) If true, all clusters in this external entity will be trusted and the `trust_allowlist` is ignored.
 * `trust_allowlist` (Optional) The list of clusters to trust. Only used when `trust_all` is `false`.
 
+##### Strategy
+
+The optional `elasticsearch.strategy` allows choosing the configuration strategy used to apply the changes. This is a fine tune setting, you do not change it unless you have a specific case where the `autodetect` do not cover your use case.
```
Comment (alaudazzi) [inline docs/resources/ec_deployment.md:385]:
```suggestion
The optional `elasticsearch.strategy` allows you to choose the configuration strategy used to apply the changes. You do not need to change this setting unless you have a specific case where the `autodetect` does not cover your use case.
```

**Review by alaudazzi — APPROVED** [review]
Left minor editing suggestions, otherwise LGTM.

**Review by tobio — APPROVED** [review]

**Pull request merged**